File: T2/Stage1/src/main/java/com/example/Stage1/Topic.java
package com.example.tarea2;

import java.util.ArrayList;
import java.util.List;

public class Topic {
    private String name;
    private List<Subscriber> subscribers = new ArrayList<>();

    public Topic(String name) {
        this.name = name;
    }

    public void subscribe(Subscriber s) {
        subscribers.add(s);
    }

    public void publish(String message) {
        for (Subscriber s : subscribers) {
            s.receive(message);
        }
    }

    public String getName() {
        return name;
    }
}

File: T3/Entrega/topic.cpp
#include "topic.h"
#include "subscriber.h"

Topic::Topic(const QString& name) : name(name) {}

QString Topic::getName() const {
    return name;
}

void Topic::addSubscriber(Subscriber* subscriber) {
    if (!subscribers.contains(subscriber)) {
        subscribers.append(subscriber);
    }
}

void Topic::removeSubscriber(Subscriber* subscriber) {
    subscribers.removeOne(subscriber);
}

void Topic::notifySubscribers(const QString& message) {
    for (Subscriber* subscriber : subscribers) {
        subscriber->update(message);
    }
}
